pandas/core/frame.py

```diff
@@ -4135,6 +4135,13 @@ def query(self, expr: str, inplace: bool = False, **kwargs):
         kwargs["target"] = None
         res = self.eval(expr, **kwargs)
 
+        if res.ndim == 1:
+            is_bool_result = is_bool_dtype(res)
+        elif res.ndim > 1:
+            is_bool_result = all(is_bool_dtype(x) for x in res.dtypes)
```

Comment:
I am not completely sure if this generalizes to higher dimensions or I can fix the condition to res.ndim == 2

Reply:
We only offer up to 2 dimensional objects in pandas so should be fine, though there also might be a more explicit way with typing to branch here. @simonjayhawkins

Reply:
you don't need any of list, you can call `common.is_bool_indexer`

Reply:
@jreback 
It correctly raises when they are not bool. But we go back to problems with tests
`pandas/tests/frame/test_query_eval.py::TestDataFrameQueryNumExprPandas::test_nested_scope`
`pandas/tests/frame/test_query_eval.py::TestDataFrameQueryPythonPandas::test_nested_scope`

The first one produces an output like the following `res` variable.
```python
import pandas as pd
from pandas.core.dtypes.common import is_bool_dtype
from pandas.core import common as com

res = pd.DataFrame([
    [False, False, False],
    [False, True, False],
    [False, False, False],
    [False, False, False],
    [False, False, False],
])

if res.ndim == 1:
    is_bool_result = is_bool_dtype(res)
elif res.ndim > 1:
    is_bool_result = all(is_bool_dtype(x) for x in res.dtypes)
new_is_bool_result = com.is_bool_indexer(res)

# is_bool_result: True
# new_is_bool: False
```

Reply:
ok when how about we use it then handle those cases. I *really* do not like re-inventing the wheel.

Reply:
The only cases is_bool_indexer handles are Series, ndarray, Index, and lists. I don't see anything in dtypes or common that handles DataFrames. I'm not sure if adding DataFrame to is_bool_indexer with its current uses is safe. Perhaps we could add `is_bool_frame`? The implementation above appears to be correct to me.

Reply:
With the new implementation that takes into account https://github.com/pandas-dev/pandas/pull/46862#discussion_r864918870 I feel we need this if/else.

In general, it seems logical to me that these cases are separated.

But if you do not agree, I want to understand a little more the proposed modification. Is this the implementation you imagined?:
```python
    if not is_bool_frame(res):
        msg = f"expr must evaluate to boolean not {res.dtypes}"
        raise ValueError(msg)
    try:
        result = self.loc[res]
    except ValueError:
        # when res is multi-dimensional loc raises, but this is sometimes a
        # valid query
        result = self[res]
```